### PR TITLE
Set transmission runs based on run angle

### LINKED ISSUE #20088: ISIS Reflectometry: Allow setting multiple transmission runs for specific angles
One of problems instrument scientists have with the current interface is that the `Transmission run(s)` setting in the `Settings` tab simply applies any specified transmission runs to all runs in the processing table. It lacks any capability to specify different transmission runs for runs with different specific angles, so they are forced to input these transmission runs in the table directly.

The `Transmission run(s)` setting needs to be changed to allow the user to input up to 2 transmission runs per angle. One possible way of doing this is to let the user input a map of angles to the transmission runs that can later be parsed, ideally in a table. For example:

| Angle | Transmission Run(s) |
|-|-|
| | 12345,12346|
| 0.5 | 22345,22346 |
| 2.3 | 32345,32346|

When we process the row, we should parse this input to apply any specified transmission runs to all runs with an angle that matches, within a certain tolerance (e.g. 100th of a degree). We should also allow a default to be specified without needing an angle lookup, e.g. by leaving the angle blank. We should only allow one default i.e. give a warning if a user enters more than one row with a blank angle.

If an angle is not specified in a row in the Runs tab, then the lookup will not be done (other than to use the default, if one was given). However, the angle will be found as part of the reduction algorithm and then populated in the table. This could possibly result in different results if you then re-run the processing, because the angle lookup will subsequently be done because it is now available in the table. To avoid confusion, we will initially disallow processing rows without an angle if per-angle settings are specified. This will not affect runs that are transferred from the Search box because they will have the angle populated. 

Ideally we should look up the angle from the log file and populate it before doing the lookup. This should be possible but is not straightforward, so needs some thought.

### PR DESCRIPTION
**Pull request by gemmaguest**
This PR adds a table to the Settings tab of the ISIS Reflectometry GUI which allows default transmission runs to be specified on a per-angle basis, so that the transmission runs can be looked up based on the angle of each run.

This replaces the existing text entry box which just allowed for one default set of transmission runs to be specified. Overall defaults can still be specified by entering one row where the angle column is blank. Due to the issues with providing the lookup angle described in the linked issue, for now, to avoid confusion, there is a restriction that per-angle OR overall defaults can be specified, but not both.

This PR also includes the following changes as part of an effort to make the underlying code more maintainable:
- Significant refactoring to cache row metadata, such as properties used in the reduction, the processed state, the reduced workspace names etc. This information was previously being recalculated in many different places whenever required but it was becoming increasingly difficult and error-prone to keep track of all of the information required to recalculate these values. These changes should make it easier going forward to e.g. change the naming convention for workspace slices.
- The `Output Notebook` option now works for all groups that are processed as non-event workspaces. Previously, if event handling was enabled but a group contained non-event workspaces, generating the notebook was being silently skipped.
- Plotting rows in event handling mode now plots the `IvsQ_binned_` workspaces rather than `IvsQ_`, to make the behaviour consistent with non-event mode.

Fixes #20088

**To test:**
Check that the `ISIS Reflectometry` and `ISIS SANS v2 experimental` GUIs still work as expected.

On the reflectometry GUI:
- Try specifying defaults with and/or without angles.
- Try overriding transmission runs in the table on the Runs tab - this should take precedence over anything on the Settings tab.
- Check that other instrument/experiment properties are still picked up ok from the Settings tab and that these also get overridden by the properties on the Runs tab.
- Try overriding properties in the Runs tab using both the normal columns (e.g. `dQ/Q`) as well as the `Options` column (e.g. `WavelengthMax` in here should override `LambdaMax` on the Settings tab).
- Check that properties and defaults are passed through ok to the preprocessing and postprocessing algorithms (i.e. `CreateTransmissionWorkspaceAuto` and `Stitch1DMany`).
- Check that the plotting buttons on the GUI work ok.
- Check that `Output Notebook` still works and that the code generated still runs ok. Check that the plots are the same as those from the GUI (although note that they are shown with error bars in the notebook and without from the GUI. Note also that there's an existing bug where the Params must be set in the stitch properties before first running the reduction because changes aren't picked up.)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by edwardb96 — CHANGES_REQUESTED** [review]
Broadly this is working. There are a few UX glitches and which could do with tidying up further along the line. There are a few code style issues outlined below, some which are easier to deal with than others...

File: qt/scientific_interfaces/ISISReflectometry/QtReflSettingsView.cpp
Comment (gemmaguest) [reply qt/scientific_interfaces/ISISReflectometry/QtReflSettingsView.cpp]: There seems to be some extra padding or something in the table so this was required to make it the correct height. I don't know if there's a better way to find out what this is.

File: qt/scientific_interfaces/ISISReflectometry/QtReflSettingsView.cpp
Comment (gemmaguest) [reply qt/scientific_interfaces/ISISReflectometry/QtReflSettingsView.cpp]: Apparently so - they are null when the GUI is first opened. I'm not sure why.

**Review by edwardb96 — APPROVED** [review]

**Comment by martyngigg** [comment]
Windows test failure is unrelated.

**Comment by gemmaguest** [comment]
I've just found a bug related to these changes so please hold off merging for now.

**Comment by gemmaguest** [comment]
To test the fix for transferring new rows: 
- Do a search in the ISIS Reflectometry GUI e.g. POLREF investigation 1520249
- Transfer several runs e.g. the first 10
- Process some rows and check you can plot them from the GUI
- Expand some groups
- Transfer some more rows to the table
- Check that:
  - the new rows get added to the correct existing group, or to a new group at the bottom of the table if their group doesn't exist yet.
  - you can still plot rows/groups that were previously reduced.
  - rows/groups that were previously highlighted green are still green (this didn't used to work so is an added benefit of this fix).
  - groups' expanded/collapsed state has not changed (again this didn't used to work)
  - if you deselect everything and hit process, rows/groups that were already reduced should not be re-processed (again this didn't used to work).

Also check that the ISIS SANS v2 experimental GUI still operates correctly

**Comment by gemmaguest** [comment]
@edwardb96 This is ready to look at again but I've left the In Progress label so it doesn't get merged yet.

**Comment by gemmaguest** [comment]
@edwardb96 Actually I have some more improvements to make so don't review yet

**Comment by gemmaguest** [comment]
@edwardb96 This is ready to look at again

**Comment by NickDraper** [comment]
Should this still have the in progress label?

**Comment by gemmaguest** [comment]
@NickDraper I left the In Progress label because I found a bug after it had been approved. Edward is going to check the latest fix and I'll remove the label when he's happy with it.

**Comment by martyngigg** [comment]
Just want to double check. Is this okay to merge now?

**Comment by edwardb96** [comment]
Not yet.

**Comment by edwardb96** [comment]
@martyngigg, @gemmaguest I have finished testing this now and it looks good to merge :+1:

**Comment by martyngigg** [comment]
Thanks.

**Pull request merged**
